### ISSUE TITLE
Provider code called in rspec testing

### ISSUE THREAD

**Issue by cdenneen** [rodjek/rspec-puppet#376]
[PUP-6067](https://tickets.puppetlabs.com/browse/PUP-6067) exposed 2 things:
1. The pip3 upgrade actually wasn't complete and caused issues in spec testing on OS X for RHEL specific code.
2. The pip, first provider, fallback was actually masking the fact that rspec-puppet is using that to mock out some facts about providers. When patch for the above bug gets merged into 4.4.1 it will be hidden again but issue is should rspec-puppet be calling that provider code. 
When compiling a catalog on master (Linux) it shouldn't be calling and looking for binaries appropriate for another platform, therefore testing on Mac shouldn't be looking for /bin/rpm, /bin/yum, /bin/mount, /bin/umount.

I think this is a long time bug exposed with the pip3 upgrade incompleteness.

**Comment by DavidS** [comment]
Hi, are you sure you are in the right repo? I do not understand what you are talking about, and PUP-6167 does not exist.

**Comment by cdenneen** [comment]
@DavidS yes, updated the link

**Comment by DavidS** [comment]
This seems to be a variation of #352 . Besides deeply intrusive mocks that disconnect most of puppet from the host system, there is not much that can be done at this point in time.

**Comment by cdenneen** [comment]
@DavidS can you explain how is this a variation of #352, that seems to be literally a puppet_gem install and test. 

This is a package resource like:

```puppet
package {'jenkins':
  ensure => '1.6.23-1',
}
```
since pip was the first provider on my system before and supported versionable this worked.
Now with 4.2.0 -> 4.4.0 pip provider isn't supporting versionable (this is being fixed in that bug ticket).
So without it rspec-puppet is looking for /bin/rpm or /bin/yum on the system to use that package provider in order to determine if it's versionable supporting or not.
I believe the facts supplied for kernel, osfamily, operatingsystem, etc in the test should be enough to determine which package provider to use vs looking for the binaries of them. As you said stubbing out these for defaultprovider isn't a long term solution because then you are left stubbing out things like /bin/mount and /bin/umount for mount {'..':} resources, and so on.

**Comment by DavidS** [comment]
> I believe the facts supplied for kernel, osfamily, operatingsystem, etc in the test should be enough to determine which package provider to use vs looking for the binaries of them.

Emphasis on "should". Puppet is not implemented this way. rspec-puppet currently has no mechnisms in place working around this globally, and it would be a massive effort to do so.

To workaround this specific puppet bugs that affect a large number of tests, I would not be opposed to a patch that sets the default package provider to something other than pip on OSX for the affected puppet versions.

**Comment by DavidS** [comment]
I was just reminded that #192 is another instance where the system running the specs leaks inexorably into the environment.

**Comment by rnelson0** [comment]
@DavidS Is there a way to mock up the provider in the rspec-puppet tests to exist? I don't follow [PUP-6067](https://tickets.puppetlabs.com/browse/PUP-6067) well enough to see how to transform that code into an rspec test to help it pass.
/cc @bastelfreak

**Comment by DavidS** [comment]
@rnelson0 the solution is in https://tickets.puppetlabs.com/browse/PUP-6067?focusedCommentId=276716&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-276716 : Michael directly stubs the default provider on the package type. 

If you look into http://rspec.info/documentation/3.4/rspec-mocks/#label-Test+Doubles you can provide a double as provider, which would allow you to replace everything at that level, at the cost of more stubbing.

Another complexity you might run into is the unholy mix of mocha and rspec-mocks. If you can, I'd recommend moving to rspec mocking syntax before starting this endeavor. See https://github.com/puppetlabs/puppetlabs_spec_helper/blob/917fd0c45a60539433c50763f7d96eb2c40a3539/lib/puppetlabs_spec_helper/puppet_spec_helper.rb#L122

**Comment by rodjek** [comment]
This should now be resolved by #498
